sbi/neural_nets/estimators/categorical_net.py

```diff
@@ -81,6 +95,34 @@ def __init__(
         self.epsilon = epsilon
         self.context_features = num_context_features
 
+    def _get_mapped_input(self, input: Tensor) -> Tensor:
+        """Map user categorical values to 0-based indices."""
+        mapped = input.clone()
+        for i, categorical_vals in enumerate(self.categorical_values_per_var):
+            # Create a mapping tensor for vectorized lookup
+            unique_vals = (
+                categorical_vals.clone()
+                .detach()
+                .to(device=input.device, dtype=input.dtype)
+            )
+            indices = torch.searchsorted(unique_vals, input[..., i].contiguous())
+            indices = indices.clamp(0, len(categorical_vals) - 1)
+            mapped[..., i] = indices
```

Comment:
this could be very costly, e.g., for many categories and when called many times during training or sampling. 
ideally, we can reduce this to an indexing operation into lookup tables that we set up during `__init__`.